Common/System/Request.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <vector>
+#include <mutex>
+#include <map>
+#include <functional>
+
+#include "Common/System/System.h"
+
+typedef std::function<void(const char *responseString, int responseValue)> RequestCallback;
+
+// Platforms often have to process requests asynchronously, on wildly different threads.
+// (Especially Android...)
+// This acts as bridge and buffer.
+class RequestManager {
+public:
+	// These requests are to be handled by platform implementations.
+	// The callback you pass in will be called on the main thread later.
+	bool MakeSystemRequest(SystemRequestType type, RequestCallback callback, const std::string &param1, const std::string &param2);
+
+	// Called by the platform implementation, when it's finished with a request.
+	void PostSystemSuccess(int requestId, const char *responseString, int responseValue = 0);
+	void PostSystemFailure(int requestId);
+
+	// This must be called every frame from the beginning of NativeUpdate().
+	// This will call the callback of any finished requests.
+	void ProcessRequests();
+
+	// Unclear if we need this...
+	void Clear();
+
+private:
+	struct PendingRequest {
+		SystemRequestType type;
+		RequestCallback callback;
+	};
+
+	std::map<int, RequestCallback> callbackMap_;
+	std::mutex callbackMutex_;
+
+	struct PendingResponse {
+		std::string responseString;
+		int responseValue;
+		RequestCallback callback;
+	};
+
+	int idCounter_ = 0;
+	std::vector<PendingResponse> pendingResponses_;
+	std::mutex responseMutex_;
+};
+
+const char *RequestTypeAsString(SystemRequestType type);
+
+extern RequestManager g_requestManager;
+
+// Wrappers for easy requests.
+// NOTE: Semantics have changed - this no longer calls the callback on cancellation.
```

Comment:
I suppose this means Osk now just hangs if you hit cancel?

-[Unknown]

Reply:
Yeah I realized recently that I can't avoid handling cancellation everywhere. I'll add it back, but using a separate callback instead - since most of the time, you don't need it.